exe/Daemon.hs

```diff
@@ -0,0 +1,515 @@
+         in writeFile ff noFollows $> noFollows
+  case A.decode (toS t) of
+    Nothing -> panic $ "Invalid daemon-follow file: could not decode " <> toS ff
+    Just fs -> pure fs
+
+writeFollowFile :: Daemon ()
+writeFollowFile = do
+  lock <- asks followFileLock
+  Chains cMap <- asks machines
+  ff <- asks followFile
+  liftIO $ withFollowFileLock lock $ do
+    ms <- CMap.nonAtomicRead cMap
+    let fs = Map.fromList $ second chainMode <$> Map.toList ms
+    writeFile ff (toS $ A.encode fs)
+
+lookupMachine :: MachineId -> Daemon (Maybe IpfsMachine)
+lookupMachine id = do
+  msVar <- asks machines
+  liftIO $ CMap.lookup id (getChains msVar)
+
+-- | Given an 'MachineId', makes sure the machine is in the cache and
+-- updated.
+checkMachineLoaded :: MachineId -> Daemon IpfsMachine
+checkMachineLoaded id = do
+  m_ <- lookupMachine id
+  case m_ of
+    Nothing -> do
+      -- In this case we have not seen the machine before so we act as
+      -- a reader.
+      m <- initAsReader id
+      writeFollowFile
+      pure m
+    Just m -> case chainMode m of
+      Writer -> pure m
+      Reader -> do
+        delta <- liftIO $ sinceLastUpdate m
+        -- If machine is half a second fresh, then return it.
+        if delta < 500
+          then pure m
+          else refreshAsReader id
+
+daemonHandler :: Env -> (Message -> Daemon ()) -> Message -> IO ()
+daemonHandler env h msg = do
+  x_ <- runDaemon env (h msg)
+  case x_ of
+    Left err -> logDaemonError err
+    Right _  -> pure ()
+
+
+
+-- Loads a machine fresh from IPFS.
+loadMachine :: ReaderOrWriter -> MachineId -> Daemon IpfsMachine
+loadMachine mode id = do
+  (idx, is) <- ipfs id $ machineInputsFrom id Nothing
+  sub <- ipfs id $ initSubscription id
+  m <- liftIO $ emptyMachine id mode sub
+  (m', _) <- addInputs is (pure idx) (const (pure ())) m
+  insertNewMachine m'
+  pure m'
+
+-- Add inputs to a cached machine.
+addInputs
+  :: [Value]
+  -- ^ Inputs to add.
+  -> Daemon MachineEntryIndex
+  -- ^ Determine the new index.
+  -> ([Value] -> Daemon ())
+  -- ^ Performed after the chain is cached.
+  -> IpfsMachine
+  -> Daemon (IpfsMachine, [Value])
+  -- ^ Returns the updated machine and the results.
+addInputs is getIdx after m =
+  case advanceChain m is of
+    Left err -> Daemon $ ExceptT $ pure $ Left $ MachineError (chainName m) (InvalidInput err)
+    Right (rs, newState) -> do
+      idx <- getIdx
+      t <- liftIO $ Time.getSystemTime
+      let news = Seq.fromList $ zip is rs
+          m' = m { chainState = newState
+                 , chainLastIndex = Just idx
+                 , chainEvalPairs = chainEvalPairs m Seq.>< news
+                 , chainLastUpdated = t
+                 }
+      after rs
+      pure (m', rs)
+
+ipfs :: MachineId -> ExceptT IpfsError IO a -> Daemon a
+ipfs id = Daemon . mapExceptT (lift . fmap (first (MachineError id . IpfsError)))
+
+-- Do some high-freq polling for a while.
```

Comment:
The polling gets bumped back down automatically?

Reply:
Yes.